components/controls/ListFooter.tsx

```diff
@@ -23,7 +23,7 @@ import { translate, translateWithParameters } from '../../helpers/l10n';
 import { formatMeasure } from '../../helpers/measures';
 import DeferredSpinner from '../ui/DeferredSpinner';
 
-export interface Props {
+export interface ListFooterProps {
```

Comment:
You don't need to export those props, you can just do `typeof ListFooter` in the test to reference the props.

Reply:
Not quite. It kinda works for the `Partial<...>` argument to the `shallowRender()` function, because it functions almost like using `any` (you can pass any prop, even if incorrect). But more importantly, it doesn't work in the type passed to `shallow<T>()`:

```tsx
function shallowRender(props: Partial<typeof ListFooter> = {}) { // <-- This (kinda) works
  return shallow<typeof ListFooter>( // <-- This doesn't
    <ListFooter count={3} loadMore={jest.fn()} total={5} {...props} />
  );
}
```

In SQ, we try to type the return of the `shallow()` function, as it allows us to use the correct type in `state()` and `props()` calls.

For example, compare this:

<img width="765" alt="1" src="https://user-images.githubusercontent.com/45544358/85690080-7fff1680-b6d3-11ea-840a-f6499214dc61.png">
<img width="1003" alt="Screenshot 2020-06-25 at 11 03 41" src="https://user-images.githubusercontent.com/45544358/85690120-8bead880-b6d3-11ea-9218-d8ec7e8d2711.png">

And this:

<img width="755" alt="2" src="https://user-images.githubusercontent.com/45544358/85690310-be94d100-b6d3-11ea-9c71-380caf12b33c.png">
<img width="1307" alt="Screenshot 2020-06-25 at 11 05 07" src="https://user-images.githubusercontent.com/45544358/85690389-d0767400-b6d3-11ea-9b37-72f0792796bf.png">

Sure, it's kind of irrelevant in this case, but it's the guidelines we try to stick with in SQ 🙂 . I'm using TS 3.8.3 here.

Of course, this is a shared project with SC. Perhaps we want to align on this between teams? I'm not against using `typeof`, but I think we should make sure we understand pros and cons of each approach.

Reply:
Wow... You are right! It doesn't work at all! I even wonder how we came to think that it works and that we should do that by default... Well I guess it's really flagrant when you do `wrapper.props().value` and we probably don't do that often for function component...

So it made me think and search a bit about it and I found this nice [`Parameters<T>` utility type](https://www.typescriptlang.org/docs/handbook/utility-types.html#parameterst), that means we could do :
```ts
function shallowRender(props: Partial<Parameters<typeof ListFooter>[0]> = {}) {
  return shallow<Parameters<typeof ListFooter>[0]>(
    <ListFooter count={3} loadMore={jest.fn()} total={5} {...props} />
  );
}
```

This works well, but it seems a bit tricky and long to write...

Reply:
Hmm, interesting. I wonder if we could create a custom type for this. Something like:

```ts
type PropsOf<T extends (...args: any) => any> = Parameters<typeof T>[0];
```

Unfortunately, the above doesn't work.

Reply:
Yeah I tried something similar but didn't manage to make it work either...
But indeed, it would be great if we can come up with a custom type like this to not have this tricky notation!

Reply:
This seems to work:
```ts
type PropsOf<T extends (...args: any) => any> = Parameters<T>[0];

function shallowRender(props: Partial<PropsOf<typeof ListFooter>> = {}) {
  return shallow<PropsOf<typeof ListFooter>>(
    <ListFooter count={3} loadMore={jest.fn()} total={5} {...props} />
  );
}
```
It's a little bit better than having to write the full thing but you still have to write the typeof...

Reply:
`Partial<PropsOf<typeof ListFooter>>` vs `ListFooterProps` ... I personally don't mind exporting the props interface and using it instead. Shorter, less error-prone, IMHO.

Reply:
Yeah I agree that it's lighter and easier... I'm just not a fan that we have to export things just for the test, and always having to wonder when reviewing if these exported props are used/imported somewhere else where it shouldn't...
But I agree that for now it seems like the "best" solution...